[Teo Zhi Hao] iP

Yttruire opened the pull request:
# DukeSupreme
> "Forget everything that you think you know" - Doctor Strange

**DukeSupreme**, or more commonly known as _Duke_, offers the following great features 👍 :
* A list of 3 type of tasks (Todo, Deadline, Event)
* Automatic saving to disk upon updating your task list
* Easy adding and deletion of these tasks

How to try out DukeSupreme:
1. Download a version of the program at [Release](https://github.com/Yttruire/ip/releases)
2. Run the .jar file
3. Try out any of the available commands: BYE, LIST, DONE, DELETE, TODO, DEADLINE, EVENT, DATE, FIND

Example of graphical task list that we provide but in command line form:
1. [ ] task
2. [x] read book (by: Aug 29 2021)
3. [ ] destroy book (at: Aug 30 2021)
4. [x] bok typo
5. [ ] book

Built on very easy to understand code. Example snippet of the `main` function:
```
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Yttruire committed:
Added Greet, Echo and Exit commands

- Changed the format of all messages sent to user
- Changed greeting message
- Added support for user input (Specifically only 3 commands)

Yttruire committed:
Changed ECHO to ADD, added LIST commands

- ECHO command has been changed to an ADD command to add to a database of strings
- LIST command added to list all stored strings

Yttruire committed:
Added Tasks and DONE command

Yttruire committed:
Added Todo, Deadline, Event, reverted ADD to ECHO

Reverted ADD to ECHO since we do not want to add generic tasks

Yttruire committed:
Updated input and expected output for automated text UI testing

Yttruire committed:
Added exceptions to handle errors and invalid commands. Removed echo

Yttruire committed:
Added Delete command

Yttruire committed:
Added save capability

Added autosave capability to a save file named "tasks.txt". The save file will be automatically loaded upon the start of program execution, and upon any update to the list of tasks, this will be automatically rewritten to the save file to update it.

Yttruire committed:
Added support for proper Date and Time representation and DATE command

Deadlines and events now represent datetime as a LocalDate object. This allows future manipulations of tasks using dates as well.

DATE command has been added to allow the query of tasks that are due/occur on the specified date.

Yttruire committed:
Merge pull request #1 from Yttruire/branch-Level-7

Added save capability

Yttruire committed:
Merge branch 'branch-level-8'

Yttruire committed:
Refactor code into more OOP

Add Ui, Storage, Parser, TaskList classes

Yttruire committed:
Refactor into packages

Divide into more sub-packages: duke.io and duke.data

Yttruire committed:
Add Unit Testing with JUnit

Add unit tests for Parser and TaskList
Add equals() override methods for TaskList, Tasks, and Command

Yttruire committed:
Add JavaDocs

Add JavaDocs for every class and public/protected method

Yttruire committed:
Move Test Cases Directory

Yttruire committed:
Add FIND command

Added parsing and handling of new FIND command.

This enables the user to get a list of all tasks whose descriptions contain a given keyword/phrase

Yttruire committed:
Check coding standard

Checked through each class for coding standard violations to identifiers, formatting, etc

Yttruire committed:
Merge branch 'branch-A-CodingStandard'

Yttruire committed:
Merge branch 'branch-Level-9'

Yttruire committed:
Fix DukeTest.java

Updated the name of the function call to test the filtering by date for a TaskList

Cyn7hius commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,149 @@
+                    LocalDate dateTime;
+                    int index;
+
+                    switch (command.getCommand()) {
```
I think this is not the correct switch statement style, the cases should be aligned with the switch statement

wowsiddanth replied:
I like how you indented the code to make the logic clear, nice! However, I did observe that your imports are in the wrong order (it should be java first , _then_ your own created ones.

Cyn7hius commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,93 @@
+            while (sc.hasNextLine()) {
+                String task = sc.nextLine();
+
+                String[] taskDetails = task.split("\\|");
```
Maybe can add a comment to explain the what each index in the taskDetails array represents?

Cyn7hius commented on src/main/java/duke/data/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    /**
+     * Contains a new TaskList containing Tasks whose description contains the given keyword/String
```
Comment is confusing, did you mean Returns a new TaskList?

Cyn7hius commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,93 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task{
```
I think a JavaDoc would be helpful here?

Cyn7hius commented on src/main/java/duke/io/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     * @return A Command object containing an identifier of the command and its arguments
+     * @throws DukeException When given input cannot be parsed successfully
+     */
+    public static Command parse(String userInput) throws DukeException {
```
I think this method is too long? Maybe can consider using helper methods to simplify the logic?

Cyn7hius reviewed (commented):
LGTM! Very good! 😉

wowsiddanth commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,149 @@
+package duke;
+
+import duke.data.Storage;
+import duke.data.TaskList;
+import duke.io.Command;
+import duke.io.Parser;
+import duke.io.Ui;
+import duke.tasks.Deadline;
+import duke.tasks.Event;
+import duke.tasks.Task;
+import duke.tasks.Todo;
+
+import java.io.IOException;
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+import java.util.Scanner;
```
I think according to the coding standards, the Java imports are below the Duke imports.

wowsiddanth commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,149 @@
+                            break;
+                        case FIND:
+                            // -1 to account for zero-indexing
+                            keyword = command.getArgs()[0];
```
If I'm not wrong, the indentation of the case should be 4 spaces to the left. Please do check as I may be wrong!

wowsiddanth commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,93 @@
+    protected File taskFile;
+
+    /**
+     * Initialises the object with a path to the save file.
```
I think the spelling should be American English,.

wowsiddanth commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,123 @@
+    protected boolean isDone;
+
+    /**
+     * Constructor accepting only description, isDone is false by default.
```
I think it should be constructs instead of constructor.

wowsiddanth reviewed (approved):
I loved your program! The logic was well-thought out and was easy to follow. Although there were a few nitpicks, it LGTM.

Yttruire committed:
Merge branch 'add-gradle-support'

Yttruire commented:
> LGTM! Very good! 😉

Hi Calvin, thank you so much for all your comments! They were all good and valuable feedback and I've incorporated them in my subsequent commits!

Yttruire commented:
> I loved your program! The logic was well-thought out and was easy to follow. Although there were a few nitpicks, it LGTM.

Hi Siddanth! Thank you for your comments as well! I did indeed miss out the coding standard violations that you have pointed out, and I've rectified them in a subsequent commit, thanks for giving me feedback!

Yttruire committed:
Add Gradle checkstyle support and refactored code according to feedback

Add Gradle checkstyle
Fix checkstyle issues
Incorporate PR review feedback

Yttruire committed:
Add GUI

Remove text-based UI
Add GUI using JavaFX

Yttruire committed:
Fix OutputFormatter JavaDocs

Yttruire committed:
Update OutputFormatter.java

Forgot to do Level-10 on separate branch

Yttruire committed:
Fix checkstyle

Fix checkstyle and prepare for gradle build to jar

Yttruire committed:
Add assertions

Add assertions to check assumptions in code.

Assertions are added in the Storage class in order to check that the data we load is in a correct/good format.

Assertions are also added in Duke to check for index out of bounds.

These are put at specific points so that if they are caught, we know that something has gone wrong with the parsing of data, either in loading from disk or there are logical errors in the data loaded.

Yttruire committed:
Improve code quality

Refactor methods that follow poor coding quality practices.

Examples of poor code quality include:
* Long methods that can be refactored into multiple smaller ones
* Methods that have deep nesting (High levels of indentation)
* Magic numbers
* No SLAP

In order to ensure greater code readability and understanding at a glance, long methods were broken down into multiple smaller ones.

This should also make it easier to read how the happy paths will execute.

Yttruire committed:
Merge pull request #3 from Yttruire/branch-A-Assertions

Add assertions

Yttruire committed:
Merge branch 'master' into branch-A-CodeQuality

Yttruire committed:
Merge pull request #4 from Yttruire/branch-A-CodeQuality

Improve code quality

Yttruire committed:
Add DoWithinPeriodTask

Add a new Task named DoWithinPeriodTask

A DoWithinPeriodTask indicates a task that contains a period, or specifically a start and end date to indicate the time period during which a task can/should be completed.

A new subclass of Task has thus be added to implement this, as this is a perfectly valid Task.

This new class can be saved to disk, formatted and marked as completed as per any other Task.

Yttruire committed:
Update build.gradle

Allows proper creation and release of working jar file

Yttruire committed:
Fix InvocationTargetException

InvocationTargetException was being unhandled when the program encounters the done and delete command without any parameters passed to it (Despite expecting one)

Yttruire committed:
Update README.md

Add description of available features as well as guide for each available command, description of each command, and examples of usage.

Yttruire committed:
Set theme jekyll-theme-slate

Yttruire committed:
Add Ui.png

Representative screenshot of the product

Yttruire committed:
Fix iP Code Quality Feedback

Fix some code quality issues based on the issues raised by the cs2103-bot

Yttruire committed:
Update README.md

Update with format and example outcomes for each command

Yttruire committed:
Update README.md

Fix to markdown
